quickwit-common/src/runtimes.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use anyhow::bail;
+use once_cell::sync::OnceCell;
+use tokio::runtime::Runtime;
+use tracing::warn;
+
+static RUNTIMES: OnceCell<HashMap<RuntimeType, tokio::runtime::Runtime>> = OnceCell::new();
+
+/// Describes which runtime an actor should run on.
+#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
+pub enum RuntimeType {
+    /// The blocking runtime runs blocking actors.
+    /// This runtime is only used as a nice thread pool with
+    /// the interface as tokio stasks.
+    ///
+    /// This runtime should not be used to run tokio
+    /// io operations.
+    ///
+    /// Tasks are allowed to block for an arbitrary amount of time.
+    Blocking,
+
+    /// The non-blocking runtime is closer to what one would expect from
+    /// a regular tokio runtime.
+    ///
+    /// Task are expect to yield within 500 microsecs.
+    NonBlocking,
+
+    /// The ingest API is a bit special in that it has its own
+    /// dedicated thread.
+    IngestApi,
+}
+
+#[derive(Debug)]
+pub struct RuntimesConfiguration {
+    num_threads_non_blocking: usize,
+    num_threads_blocking: usize,
+}
+
+impl RuntimesConfiguration {
+    fn light_for_tests() -> RuntimesConfiguration {
+        RuntimesConfiguration {
+            num_threads_blocking: 1,
+            num_threads_non_blocking: 1,
+        }
+    }
+
+    pub fn with_num_cpus(num_cpus: usize) -> Self {
+        let num_threads_non_blocking = if num_cpus > 6 { 2 } else { 1 };
+        let num_threads_blocking = (num_cpus - num_threads_non_blocking).max(1);
```

Comment:
The rationale here is that we want to give the maximum number of threads to the CPU-consuming actors: Indexer, Merger, Packager.
You can add a comment maybe to justify the default values?